### PR TITLE
[WIP][MARXAN-1122]adds-ACL-scenarios-controller

### PR DESCRIPTION
**Pull request by rubvalave**
-Adds ACL to every endpoint in the scenarios-controller.
-Fixes some missing ImplementsACL decorator not present in projects-listing-controller.
-Various additions to ACL-utils.

### CODE REVIEW

File: api/apps/api/src/modules/scenarios/scenarios.controller.ts [hunk 1]
```diff
@@ -213,9 +234,20 @@ export class ScenariosController {
   @Get(':id/planning-units/tiles/:z/:x/:y.mvt')
   async proxyProtectedAreaTile(
     @Req() request: Request,
+    @Req() req: RequestWithAuthenticatedUser,
```
Comment (tiagojsag) [inline api/apps/api/src/modules/scenarios/scenarios.controller.ts:237]: Do we really need the request twice here? The arg you added extends Request, so is there any reason why it can't simply replace the existing Request?

File: api/apps/api/src/modules/scenarios/scenarios.controller.ts [hunk 2]
```diff
@@ -213,9 +234,20 @@ export class ScenariosController {
...
   ) {
-    return this.proxyService.proxyTileRequest(request, response);
+    /* Due to the usage of proxyService in other modules
+    the ACL control for this endpoint is placed in the controller */
+    if (
+      !(await this.scenarioAclService.canViewScenario(req.user.id, scenarioId))
+    ) {
+      throw new ForbiddenException();
+    }
+    const result = await this.proxyService.proxyTileRequest(request, response);
```
Comment (tiagojsag) [inline api/apps/api/src/modules/scenarios/scenarios.controller.ts:248]:
```suggestion
    return this.proxyService.proxyTileRequest(request, response);
```

File: api/apps/api/src/modules/scenarios/scenarios.service.ts [hunk 3]
```diff
@@ -319,49 +476,86 @@ export class ScenariosService {
   async getOneSolution(
     scenarioId: string,
     runId: string,
+    userId: string,
     _fetchSpecification: FetchSpecification,
-  ) {
+  ): Promise<Either<typeof forbiddenError, ScenariosOutputResultsApiEntity>> {
     await this.assertScenario(scenarioId);
-    // TODO permissions guard
-    // TODO runId is part of scenarioId
```
Comment (tiagojsag) [inline api/apps/api/src/modules/scenarios/scenarios.service.ts:326]: Has this been handled?

File: api/apps/api/src/modules/scenarios/scenarios.service.ts [hunk 4]
```diff
@@ -408,67 +606,133 @@ export class ScenariosService {
...
+  async createSpecification(
+    scenarioId: string,
+    userId: string,
+    dto: CreateGeoFeatureSetDTO,
+  ): Promise<Either<typeof forbiddenError | typeof internalError, any>> {
     const scenario = await this.assertScenario(scenarioId);
+    if (!(await this.scenarioAclService.canEditScenario(userId, scenarioId))) {
+      return left(forbiddenError);
+    }
     return await this.specificationService.submit(
```
Comment (tiagojsag) [inline api/apps/api/src/modules/scenarios/scenarios.service.ts:671]: right() missing here?

File: api/apps/api/src/modules/scenarios/scenarios.service.ts [hunk 5]
```diff
@@ -408,67 +606,133 @@ export class ScenariosService {
...
+    Either<
+      LastUpdatedSpecificationError | typeof forbiddenError,
+      CreateGeoFeatureSetDTO
+    >
+  > {
     const scenario = await this.assertScenario(scenarioId);
+    if (!(await this.scenarioAclService.canViewScenario(userId, scenarioId))) {
+      return left(forbiddenError);
+    }
     return await this.specificationService.getLastUpdatedFor(
```
Comment (tiagojsag) [inline api/apps/api/src/modules/scenarios/scenarios.service.ts:691]: and here?

**Review by tiagojsag — CHANGES_REQUESTED** [review]
- Seems like you have a few rigth() calls missing. I highlighted a couple, but there seem to be more
- lots of tests missing - those missing right() calls should have been caught by the tests you didn't add
- return await is more often than not an anti pattern

**Review by hotzevzl — COMMENTED** [review]
thanks @rubvalave. I am not too convinced with the new `findAllPaginatedAcl()` method, as this would leave the base class' own `findAllPaginated()` open to accidental use without access control checks.

One possible way to handle this would be what I sketched quickly in https://github.com/Vizzuality/marxan-cloud/tree/feat/MARXAN-1122-scenario-acl-application%2Bandrea-suggestions.

That's still not ideal because the base class then becomes coupled with ACL concerns, which I'd say could be ok because it's the marxan backend's own wrapper for `BaseService`, but one could also argue otherwise.

Another possible way would be to make `BaseService.findAllPaginated()` a private method for internal use, renaming it for example to `BaseService._findAllPaginated()` as with other methods that we wouldn't want consumers to rely on, and implement `findAllPaginated()` as per your implementation, but with the leaner name (without the `Acl` part).

Either way, there could be arguments for and against, so I'm just bringing forward some suggestions.

**Comment by hotzevzl** [comment]
> thanks @rubvalave. I am not too convinced with the new `findAllPaginatedAcl()` method, as this would leave the base class' own `findAllPaginated()` open to accidental use without access control checks.
> 
> One possible way to handle this would be what I sketched quickly in https://github.com/Vizzuality/marxan-cloud/tree/feat/MARXAN-1122-scenario-acl-application%2Bandrea-suggestions.

so quickly in fact that it's obviously incomplete and doesn't even transpile, but just to give a rough idea 😇

File: api/apps/api/src/modules/scenarios/scenarios.controller.ts (hunk 1 above)
Comment (hotzevzl) [reply api/apps/api/src/modules/scenarios/scenarios.controller.ts:237]: my fault (mindlessly accepting IDE suggestions) - `RequestWithAuthenticatedUser` is a remnant of past projects, and I brought it over importing the wrong `Request`, so fixing `RequestWithAuthenticatedUser` to extend `Request` from Express is the correct way to handle this.

File: api/apps/api/src/modules/scenarios/scenarios.service.ts (hunk 3 above)
Comment (rubvalave) [reply api/apps/api/src/modules/scenarios/scenarios.service.ts:326]: Permissions yes, but I overlooked that I removed the latter part so I added it again.

File: api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts [hunk 6]
```diff
@@ -1,19 +1,61 @@
...
-describe(`When scenario has PUs with cost and lock status`, () => {
-  beforeAll(async () => {
-    await world.GivenScenarioWithPuAndLocks();
+describe(`As owner, when scenario has PUs with cost and lock status`, () => {
+  it(`returns relevant data for pu.dat`, async () => {
+    const result = await fixtures.WhenGettingMarxanDataAsOwner();
+    const [headers, ...costAndStatus] = result.split('\n');
+
+    expect(headers).toEqual('id\tcost\tstatus');
+    expect(costAndStatus).toMatchInlineSnapshot(`
```
Comment (AaronPerezPerez) [inline api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts:18]: Wouldn't it be better to use .toEqual() ?
It seems clearer IMHO
```suggestion
    expect(costAndStatus).toEqual([
        "0	200	0",
        "1	400	2",
        "2	600	3",
        "3	800	0",
      ])
```

File: api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts [hunk 7]
```diff
@@ -54,6 +96,62 @@ describe(`When scenario has PUs with cost and lock status`, () => {
...
 
+describe(`As viewer, when scenario has PUs with cost and lock status`, () => {
+  it(`returns relevant data for pu.dat`, async () => {
+    await fixtures.GivenViewerWasAddedToScenario();
+    const result = await fixtures.WhenGettingMarxanDataAsViewer();
+    const [headers, ...costAndStatus] = result.split('\n');
+
+    expect(headers).toEqual('id\tcost\tstatus');
+    expect(costAndStatus).toMatchInlineSnapshot(`
+      Array [
```
Comment (AaronPerezPerez) [inline api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts:107]: Same as before

File: api/apps/api/test/scenario-protected-areas/fixtures.ts [hunk 8]
```diff
@@ -124,6 +167,17 @@ export const getFixtures = async () => {
         selected: true,
       });
     },
+    ThenItDoesNotContainsSelectedCustomArea: async (
+      response: any[],
```
Comment (AaronPerezPerez) [inline api/apps/api/test/scenario-protected-areas/fixtures.ts:171]: Can we add proper typing to this param?

File: api/apps/api/test/scenario-protected-areas/fixtures.ts [hunk 9]
```diff
@@ -165,6 +262,17 @@ export const getFixtures = async () => {
         selected: true,
       });
     },
+    ThenItDoesNotContainsSelectedGlobalArea: async (
+      response: any[],
```
Comment (AaronPerezPerez) [inline api/apps/api/test/scenario-protected-areas/fixtures.ts:266]: Same as before

File: api/apps/api/test/scenario-solutions/get-all-solutions.fixtures.ts [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    WhenGettingSolutionsAsContributor: async () =>
+      request(app.getHttpServer())
+        .get(`/api/v1/scenarios/${scenarioId}/marxan/solutions`)
+        .set('Authorization', `Bearer ${contributorToken}`),
+    WhenGettingSolutionsAsViewer: async () =>
+      request(app.getHttpServer())
+        .get(`/api/v1/scenarios/${scenarioId}/marxan/solutions`)
+        .set('Authorization', `Bearer ${viewerToken}`),
+    ThenSolutionsShouldBeResolved: async (response: request.Response) => {
+      expect(response.body.meta).toMatchInlineSnapshot(`
```
Comment (AaronPerezPerez) [inline api/apps/api/test/scenario-solutions/get-all-solutions.fixtures.ts:84]: Again, I prefer using `.toEqual()`, but above all I prefer that we use the same method (any of them) everywhere
```suggestion
      expect(response.body.meta).toEqual({
        "page": 1,
        "size": 25,
        "totalItems": 1,
        "totalPages": 1,
      })
```

File: api/apps/api/test/scenario-solutions/get-all-solutions.fixtures.ts [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+      Object {
+        "page": 1,
+        "size": 25,
+        "totalItems": 1,
+        "totalPages": 1,
+      }
+    `);
+      expect(response.body.data.length).toEqual(1);
+      expect(response.body.data[0].attributes).toMatchInlineSnapshot(
```
Comment (AaronPerezPerez) [inline api/apps/api/test/scenario-solutions/get-all-solutions.fixtures.ts:93]: Same as before

**Review by AaronPerezPerez — CHANGES_REQUESTED** [review]

File: api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts (hunk 6 above)
Comment (rubvalave) [reply api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts:18]: To be honest, for most of those that use `MatchInLineSnapshot`, I just added variants without changing much the original tests done by Kamil. I guess this is a matter of preference, and yeah maybe it will be better if we use the same method for every test.

Comment (tiagojsag) [reply api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts:18]: I second @AaronPerezPerez recommendation.

**Review by tiagojsag — APPROVED** [review]
TL;DR, so I approve, but if you are up for changing those `toMatchInlineSnapshot` calls, great

**Review by hotzevzl — APPROVED** [review]

**Review by AaronPerezPerez — APPROVED** [review]

File: api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts (hunk 6 above)
Comment (hotzevzl) [reply api/apps/api/test/scenario-input-files/scenario-cost-surface/scenario-cost-surface.e2e-spec.ts:18]: @rubvalave though I'm fine with either, I think the _matter of preference_ bit is mostly about workflow, not so much code style per se.

Snapshots (and the inline variant used throughout these tests) can be useful if wanting to reap some free labour from Jest, as they can be updated via `jest --updateSnapshot ----testNamePattern <pattern>` rather than having to manually edit the expected values for `toEqual()`/`toStrictEqual()` (the latter being what we likely want in most cases, tin-foil hat on).

I'd say that for small objects there is no practical difference. When checking asserting equality of complex objects such as a, say GeoJSON document, then snapshots can help substantively.
